python/lsst/daf/butler/_dataset_provenance.py

```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+
+        :id: UUID of the given dataset.
+        :run: Run of the given dataset.
+        :datasettype: Dataset type of the given dataset.
+        :dataid x: An entry for each required dimension, "x", in the data ID.
+
+        Each input dataset will have the ``id``, ``run``, and ``datasettype``
+        keys as defined above (but no ``dataid`` key) with an ``input N``
+        prefix where ``N`` starts counting at 0.
+
+        The quantum ID, if present, will use key ``quantum``.
+
+        Examples
+        --------
+        >>> provenance.to_flat_dict(
+        ...     ref, prefix="lsst.butler", sep=".", simple_types=True
+        ... )
+        {
+            "lsst.butler.id": "ae0fa83d-cc89-41dd-9680-f97ede49f01e",
+            "lsst.butler.run": "test_run",
+            "lsst.butler.datasettype": "data",
+            "lsst.butler.dataid.detector": 10,
+            "lsst.butler.dataid.instrument": "LSSTCam",
+            "lsst.butler.quantum": "d93a735b-08f0-477d-bc95-2cc32d6d898b",
+            "lsst.butler.input.0.id": "3dfd7ba5-5e35-4565-9d87-4b33880ed06c",
+            "lsst.butler.input.0.run": "other_run",
+            "lsst.butler.input.0.datasettype": "astropy_parquet",
+            "lsst.butler.input.1.id": "7a99f6e9-4035-3d68-842e-58ecce1dc935",
+            "lsst.butler.input.1.run": "other_run",
+            "lsst.butler.input.1.datasettype": "astropy_parquet",
+        }
+
+        Raises
+        ------
+        ValueError
+            Raised if the separator is not a single character.
+        """
+
+        def _make_key(*keys: str | int) -> str:
+            """Make the key in the correct form with simpler API."""
+            return self._make_provenance_key(prefix, sep, *keys)
+
+        prov: dict[str, int | float | str | bool | uuid.UUID | DataIdValue] = {}
+        if ref is not None:
+            prov[_make_key("id")] = ref.id if not simple_types else str(ref.id)
+            prov[_make_key("run")] = ref.run
+            prov[_make_key("datasettype")] = ref.datasetType.name
+            for k, v in sorted(ref.dataId.required.items()):
+                prov[_make_key("dataid", k)] = v
+
+        if self.quantum_id is not None:
+            prov[_make_key("quantum")] = self.quantum_id if not simple_types else str(self.quantum_id)
+
+        for i, input in enumerate(self.inputs):
+            prov[_make_key("input", i, "id")] = input.id if not simple_types else str(input.id)
+            if input.run is not None:  # for mypy
+                prov[_make_key("input", i, "run")] = input.run
+            if input.datasetType is not None:  # for mypy
+                prov[_make_key("input", i, "datasettype")] = input.datasetType.name
+
+            if input.id in self.extras:
+                for xk, xv in self.extras[input.id].items():
+                    if simple_types and not isinstance(xv, str | float | int | bool):
+                        xv = str(xv)
+                    prov[_make_key("input", i, xk)] = xv
```

Comment:
Could `self.extras` contain keys which conflict with other keys, e.g. `datasettype`?

Reply:
Sigh. Yes. Hmm. This is why I ask you for reviews... I think the fix is to validate add_extra_provenance and disallow run, id, and datasettype as keys.

Reply:
Validation added.